cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/AnomalyDetector.java

```diff
@@ -418,34 +434,40 @@ private void fixAnomalyInProgress(AnomalyType anomalyType) throws Exception {
           LOG.info("Skip fixing anomaly {}, because anomaly detector is shutting down.", _anomalyInProgress);
         } else {
           boolean isReadyToFix = isAnomalyInProgressReadyToFix(anomalyType);
-          if (isReadyToFix) {
-            LOG.info("Fixing anomaly {}", _anomalyInProgress);
-            boolean startedSuccessfully = false;
-            String anomalyId = _anomalyInProgress.anomalyId();
-            try {
+          boolean startedSuccessfully = false;
+          String anomalyId = _anomalyInProgress.anomalyId();
+          try {
+            if (isReadyToFix) {
+              LOG.info("Fixing anomaly {}", _anomalyInProgress);
               startedSuccessfully = _anomalyInProgress.fix();
               String optimizationResult = optimizationResult(anomalyType);
               _anomalyLoggerExecutor.submit(() -> logSelfHealingOperation(anomalyId, null, optimizationResult));
-            } catch (OptimizationFailureException ofe) {
-              _anomalyLoggerExecutor.submit(() -> logSelfHealingOperation(anomalyId, ofe, null));
-              throw ofe;
-            } finally {
-              _anomalyDetectorState.onAnomalyHandle(_anomalyInProgress, startedSuccessfully ? AnomalyState.Status.FIX_STARTED
-                                                                                            : AnomalyState.Status.FIX_FAILED_TO_START);
-              if (startedSuccessfully) {
-                _anomalyDetectorState.incrementNumSelfHealingStarted();
-                LOG.info("[{}] Self-healing started successfully.", anomalyId);
-              } else {
-                LOG.warn("[{}] Self-healing failed to start.", anomalyId);
-              }
             }
+          } catch (OptimizationFailureException ofe) {
+            _anomalyLoggerExecutor.submit(() -> logSelfHealingOperation(anomalyId, ofe, null));
+            throw ofe;
+          } finally {
+            handlePostFixAnomaly(isReadyToFix, startedSuccessfully, anomalyId);
```

Comment:
If we get an `OptimizationFailureException`(due to hard goal violation) when fixing a `BrokerFailure` here, this `handlePostFixAnomaly` will be called and we schedule a `detectBrokerFailures()` with 0ms delay. 

This  `detectBrokerFailures()` is called with `skipReportingIfNotUpdated` is set to false,  so a new `BrokerFailure` will be reported immediately.

Then even if `SelfHealingNotifier` will not send a new alert(which is good) but handler will try to fix and fail again. Looks like it forms a loop?

Reply:
Thanks -- addressed the issue!